### PR TITLE
Add API reference documentation for configurations and controllers (close #428)

### LINKED ISSUE #428: Add API reference documentation for configurations and controllers
Add Javadoc comments to document the new API and deprecate the old API indicating the alternative methods.

### PR DESCRIPTION
**Pull request by AlexBenny**
In this PR I've documented most of new public API about configuration of the tracker and its controlling at runtime.
Also I've deprecated much of the old API redirecting to the new API.

### CODE REVIEW

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SessionConfiguration.java [hunk 1]
```diff
@@ -6,39 +6,79 @@
 import com.snowplowanalytics.snowplow.internal.session.SessionConfigurationInterface;
 import com.snowplowanalytics.snowplow.util.TimeMeasure;
 
+/**
+ * This class allows the tracker configuration from the session perspective.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SessionConfiguration.java:10]:
```suggestion
 * This class represents the configuration from of the applications session.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SessionConfiguration.java [hunk 2]
```diff
@@ -6,39 +6,79 @@
...
+ * The SessionConfiguration can be used to setup the behaviour of sessions.
+ *
+ * A session is a context which is appended to each event sent.
+ * The values it brings can change based on:
+ * - the timeout set for the inactivity of app when in foreground;
+ * - the timeout set for the inactivity of app when in background.
+ *
+ * Session data is maintained for the life of the application being installed on a device.
+ * Essentially will update if it is not accessed within a configurable timeout.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SessionConfiguration.java:19]:
```suggestion
 * A new session will be created if the session information is not accessed within a configurable timeout.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java [hunk 3]
```diff
@@ -5,87 +5,159 @@
 
 import com.snowplowanalytics.snowplow.util.Size;
 
+/**
+ * This class allows the tracker configuration from the subject perspective.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java:9]:
```suggestion
 * This class represents the configuration of the subject.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java [hunk 4]
```diff
@@ -5,87 +5,159 @@
...
+     */
     @NonNull
     public SubjectConfiguration networkUserId(@Nullable String networkUserId) {
         this.networkUserId = networkUserId;
         return this;
     }
 
+    /**
+     * The domain UserID.
+     * @apiNote It's not generated by the tracker, it needs to be filled by the developer on tracker instrumentation.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java:92]:
```suggestion
     * @apiNote It's not generated by the tracker, it needs to be filled by the developer when instrumenting the tracker.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java [hunk 5]
```diff
@@ -5,87 +5,159 @@
...
+     */
     @NonNull
     public SubjectConfiguration userId(@Nullable String userId) {
         this.userId = userId;
         return this;
     }
 
+    /**
+     * The network UserID.
+     * @apiNote It's not generated by the tracker, it needs to be filled by the developer on tracker instrumentation.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java:82]:
```suggestion
     * @apiNote It's not generated by the tracker, it needs to be filled by the developer when instrumenting the tracker.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java [hunk 6]
```diff
@@ -5,87 +5,159 @@
...
+     */
     @NonNull
     public SubjectConfiguration domainUserId(@Nullable String domainUserId) {
         this.domainUserId = domainUserId;
         return this;
     }
 
+    /**
+     * The user-agent.
+     * @apiNote It's not generated by the tracker, it needs to be filled by the developer on tracker instrumentation.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java:102]:
```suggestion
     * @apiNote It's not generated by the tracker, it needs to be filled by the developer when instrumenting the tracker.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java [hunk 7]
```diff
@@ -5,87 +5,159 @@
...
+     */
     @NonNull
     public SubjectConfiguration useragent(@Nullable String useragent) {
         this.useragent = useragent;
         return this;
     }
 
+    /**
+     * The IP address.
+     * @apiNote It's not generated by the tracker, it needs to be filled by the developer on tracker instrumentation.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java:112]:
```suggestion
     * @apiNote It's not generated by the tracker, it needs to be filled by the developer when instrumenting the tracker.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java [hunk 8]
```diff
@@ -5,87 +5,159 @@
...
+     */
     @NonNull
     public SubjectConfiguration screenResolution(@Nullable Size screenResolution) {
         this.screenResolution = screenResolution;
         return this;
     }
 
+    /**
+     * The screen viewport.
+     * @apiNote It's not generated by the tracker, it needs to be filled by the developer on tracker instrumentation.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java:149]:
```suggestion
     * @apiNote It's not generated by the tracker, it needs to be filled by the developer when instrumenting the tracker.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java [hunk 9]
```diff
@@ -5,87 +5,159 @@
...
+     */
     @NonNull
     public SubjectConfiguration screenViewPort(@Nullable Size screenViewPort) {
         this.screenViewPort = screenViewPort;
         return this;
     }
 
+    /**
+     * The color depth.
+     * @apiNote It's not generated by the tracker, it needs to be filled by the developer on tracker instrumentation.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/SubjectConfiguration.java:159]:
```suggestion
     * @apiNote It's not generated by the tracker, it needs to be filled by the developer when instrumenting the tracker.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/TrackerConfiguration.java [hunk 10]
```diff
@@ -3,34 +3,87 @@
...
 import androidx.annotation.Nullable;
 
+import com.snowplowanalytics.snowplow.emitter.BufferOption;
 import com.snowplowanalytics.snowplow.internal.tracker.TrackerConfigurationInterface;
 import com.snowplowanalytics.snowplow.tracker.DevicePlatform;
 import com.snowplowanalytics.snowplow.tracker.LoggerDelegate;
 import com.snowplowanalytics.snowplow.tracker.LogLevel;
 
+/**
+ * This class allows the tracker configuration from the tracking perspective.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/TrackerConfiguration.java:13]:
```suggestion
 * This class represents the configuration of the tracker and the core tracker properties.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/TrackerConfiguration.java [hunk 11]
```diff
@@ -214,90 +286,139 @@ public TrackerConfiguration(@NonNull String appId) {
...
     @NonNull
     public TrackerConfiguration platformContext(boolean platformContext) {
         this.platformContext = platformContext;
         return this;
     }
 
+    /**
+     * Whether geo-location context is sent with all the tracked events.
+     *
+     * @apiNote Requires Location permissions accordingly to the requirements of the various
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/TrackerConfiguration.java:355]:
```suggestion
     * @apiNote Requires Location permissions as per the requirements of the various
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/controller/SessionController.java [hunk 12]
```diff
@@ -6,17 +6,55 @@
 
 public interface SessionController extends SessionConfigurationInterface {
 
+    /**
+     * The session index.
+     * A progressive number helping to order the sequence of sessions.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/controller/SessionController.java:11]:
```suggestion
     * A increasing number which helps to order the sequence of sessions.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/controller/SessionController.java [hunk 13]
```diff
@@ -6,17 +6,55 @@
...
+    /**
+     * The session identifier.
+     * A unique identifier which is used to identify the session.
+     */
     @NonNull
     String getSessionId();
+
+    /**
+     * The session user identifier.
+     * It identifies this app installation and it doesn't change along all the life of the app.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/controller/SessionController.java:24]:
```suggestion
     * It identifies this app installation and it doesn't change for the life of the app.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/emitter/storage/EventStoreHelper.java [hunk 14]
```diff
@@ -57,6 +57,17 @@
     // Prevents multiple instances being created and avoids memory leaks.
     private static HashMap<String, EventStoreHelper> instances = new HashMap<>();
 
+    /**
+     * IMPORTANT: This method is for internal use only. It can change signature and behaviour in any
+     * next tracker release.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/emitter/storage/EventStoreHelper.java:62]:
```suggestion
     * IMPORTANT:  This method is for internal use only. It's signature and behaviour might change in any
     * future tracker release.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/emitter/storage/EventStoreHelper.java [hunk 15]
```diff
@@ -57,6 +57,17 @@
...
+     *
+     * It clear all the EventStores not associated at any of the namespaces passed as parameter.
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/emitter/storage/EventStoreHelper.java:64]:
```suggestion
     * Clears all the EventStores not associated at any of the namespaces passed as parameter.
```

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/TrackerConfigurationInterface.java [hunk 16]
```diff
@@ -9,67 +9,165 @@
...
+     *
+     * @apiNote Requires Location permissions accordingly to the requirements of the various
+     * Android versions. Otherwise the whole context is skipped.
+     */
     boolean isGeoLocationContext();
 
+    /**
+     * Whether geo-location context is sent with all the tracked events.
+     *
+     * @apiNote Requires Location permissions accordingly to the requirements of the various
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/TrackerConfigurationInterface.java:97]: See earlier comment

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/TrackerConfigurationInterface.java [hunk 17]
```diff
@@ -9,67 +9,165 @@
...
 
+    /**
+     * Whether mobile/platform context is sent with all the tracked events.
+     */
     void setPlatformContext(boolean platformContext);
 
+    /**
+     * Whether geo-location context is sent with all the tracked events.
+     *
+     * @apiNote Requires Location permissions accordingly to the requirements of the various
```
Comment (paulboocock) [inline snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/TrackerConfigurationInterface.java:89]: See earlier comment

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/NetworkConfiguration.java [hunk 18]
```diff
@@ -20,23 +26,49 @@
...
+     * @see #okHttpClient(OkHttpClient)
+     */
+    @Nullable
+    public OkHttpClient okHttpClient;
+
     // Constructors
 
+    /**
+     * @param endpoint URL of the collector that is going to receive the events tracked by the tracker.
+     *                 The URL can include the schema/protocol (e.g.: `http://collector-url.com`).
```
[the remark this answers is not in the document]
Comment (AlexBenny) [reply snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/NetworkConfiguration.java:57]: Android Studio thinks adding spaces is more readable. However, the rendering doesn't consider them.

**Pull request merged**